Include nftables information in Agent supportbundle

molegit9 opened the pull request:
Fixes: https://github.com/antrea-io/antrea/issues/7546
Modification location: pkg/support/dump_others.go

I corrected it because the diagnostic file did not contain the information of 'NFTables'.
' 'pkg/support/dump_others.go' has been added with the function 'dumpNFTables' to allow information to be included in the diagnosis

molegit9 commented:
/test-all

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +63,40 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+	}
+
+	if d.v4Enabled && c.IPv4 != nil {
+		output, err := c.IPv4.DumpTable(context.TODO(), "antrea")
```
I didn't find the method `DumpTable` in the interface from nftables lib https://github.com/hongliangl/knftables/blob/e4307300abb50f2d9e7a492c474fe046d777f87f/nftables.go#L30.

Please use the appropriate methods to dump the table.

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +63,40 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+		return nil
+	}
+
+	if d.v4Enabled && c.IPv4 != nil {
```
```suggestion
	if c.IPv4 != nil {
```

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +63,40 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+		}
+	}
+
+	if d.v6Enabled && c.IPv6 != nil {
```
```suggestion
	if c.IPv6 != nil {
```

hongliangl reviewed (commented):
I would save the results of `table ip antrea` and `table ip6 antrea` to one file `nftables`, just like file `iptables`.

Additionally, could you add the unit tests for this?

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +64,49 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+		if err != nil {
+			klog.ErrorS(err, "Failed to dump nftables table 'ip antrea'")
+			return err
+		}
```
```suggestion
		if err != nil {
			return fmt.Errorf("failed to dump nftables table 'ip antrea': %w", err)
		}
```

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +64,49 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+		if err != nil {
+			klog.ErrorS(err, "Failed to dump nftables table 'ip6 antrea'")
+			return err
+		}
```
```suggestion
		if err != nil {
			return fmt.Errorf("failed to dump nftables table 'ip6 antrea': %w", err)
		}
```

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +64,49 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+	fileName := "nftables"
+	if err := writeFile(d.fs, filepath.Join(basedir, fileName), fileName, data.Bytes()); err != nil {
+		klog.ErrorS(err, "Failed to write nftables file", "file", fileName)
+		return err
```
```suggestion
		return fmt.Errorf("failed to write nftables file: %w", err)
```

hongliangl commented on pkg/support/dump_others_test.go:
```diff
@@ -4,7 +4,7 @@
 // You may obtain a copy of the License at
 //
-//     http://www.apache.org/licenses/LICENSE-2.0
+// 		http://www.apache.org/licenses/LICENSE-2.0
```
Revert this.

hongliangl reviewed (commented):
LGTM overall from me, only a nit in PR description:

`issue : #7546` -> `Fixes`: #7546

For git commit and PR titles, how about removing the prefix `Fix(#7546):`?

Defer to @antoninbas @luolanzone to do the final approval.

molegit9 commented:
> Please resolve the failed checks :
> 
> * Go/Golangci-lint
> * Go/Unit test
> * DCO

Go/Golangci-lint I reordered the import statements using goimports -w -local antrea.io/antrea pkg/support/dump_others.go.

Go/Unit test I modified the code to not write data when there is no output (empty string), like so:

dump_others.go
line 82, 93
if len(output) > 0 {
	data.Write(output)
	data.WriteString("\n")
}

dump_others_test.go
line 174
Set 'fakeExecutor := &testingexec.FakeExec{}' to ensure it returns the proper fake results.

DCO I corrected the Signed-off-by line from just a name to the full email address.

molegit9 commented:
In agent.go, I made it check for nftables support once at agent startup using nftclient.New(), save it to nftablesSupported, and added it as an argument to the functions.

And I changed the existing LookPath check method to check via nftablesSupported.

There were no problems when I ran make golangci and make .linux-test-unit locally, but I don't know if other problems will occur as I lack knowledge.

Please review the code, and if you find it strange, I will re-upload a version with only the log removal that the reviewer requested.

hongliangl commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +65,50 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
+		klog.ErrorS(err, "'nft' command not found")
+		return nil
+	}
+
```
How about this:

1. Only initialize the wrapper library `Client` to check if nftables is available. This is because the initialization process does the following checks. See the link https://github.com/antrea-io/antrea/blob/b36df946c1639ab8235b43cc602468734875cb69/pkg/agent/util/nftables/nftables.go#L67

```
	// knftables.New validates:
	//  - nft binary is available
	//  - sufficient permissions
	//  - kernel version compatibility
	//  - "nft destroy" support
```

If an error is got when initializing, just log something that nftables is not available and skip dumping nftables.

2. Use the `d.executor.Command("nft", "list", "table", "ip", "antrea").CombinedOutput()` to dump the table.

The first step might be confusing, and we need to document it well to state that this step is to verify the availability of nftables.

molegit9 replied:
Hi @antoninbas and @hongliangl,

Just a gentle ping regarding the updates based on your previous discussion.

I've refactored the PR according to the approach you mentioned:

Startup Check: Added nftclient.New in agent.go to verify kernel compatibility during initialization.

Dump Execution: Kept d.executor.Command("nft", ...) for the actual dump since the library doesn't support a full dump.

Log Cleanup: Removed the "No Antrea nftables rules found to dump" log message.

I believe this reflects all the points you raised.
When you have a moment, could you please take a look?

Thanks!

hongliangl commented on cmd/antrea-agent/agent.go:
```diff
@@ -326,11 +329,21 @@ func run(o *Options) error {
+		klog.InfoS("nftables is supported on this Node")
+		nftablesSupported = true
+	}
+
```
We can just move the logic L338-L346 to function `dumpNFTables` in dump_others.go, and we don't need the changes of introducing extra parameter `nftablesSupported`, simplifying the PR.

molegit9 commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +66,48 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
 
+func (d *agentDumper) dumpNFTables(basedir string) error {
+	if _, err := nftclient.New(d.v4Enabled, d.v6Enabled); err != nil {
+		klog.V(4).InfoS("Skipping nftables dump because it is not supported on this Node", "err", err)
```
Thank you for the review!
'klog.V(4).InfoS' -> 'ErrorS'
I modified it.

hongliangl reviewed (dismissed):
LGTM, defer to @antoninbas and @luolanzone to do final approvals.

molegit9 commented on pkg/support/dump_others.go:
```diff
@@ -18,15 +18,24 @@
+// newNFTablesClient is defined as a variable to allow mocking in unit tests.
+var newNFTablesClient = func(v4Enabled, v6Enabled bool) (interface{}, error) {
+	return nftclient.New(v4Enabled, v6Enabled)
+}
```
Thanks for the review. I've updated the PR based on your feedback.

antoninbas commented on pkg/support/dump_others.go:
```diff
@@ -18,15 +18,38 @@
+// They initialize the client once to verify support, but the returned clients are not used.
+var nftablesIPv4Supported = sync.OnceValue(func() bool {
+	if _, err := nftclient.New(true, false); err != nil {
+		klog.ErrorS(err, "nftables IPv4 not supported on this Node")
```
I would switch it to `klog.InfoS` - it may be a pretty harmless error in some cases - even if we don't end up logging the actual `err`

antoninbas reviewed (commented):
only minor comments left, otherwise lgtm

antoninbas commented on pkg/support/dump_others.go:
```diff
@@ -59,6 +85,44 @@ func (d *agentDumper) dumpIPTables(basedir string) error {
 }
 
+func (d *agentDumper) dumpNFTables(basedir string) error {
+
```
remove empty line

hongliangl commented:
> @molegit9 The e2e tests are failing. There wasn't much info in the test logs, but I saw this log message in one of the Agents' logs:
> 
> ```
> E1203 00:13:57.544016      13 rest.go:161] "Error when collecting supportBundle" err="failed to dump nftables table 'ip antrea': exit status 1"
> ```
> 
> Maybe `nftablesIPv4Supported` is incomplete? cc @hongliangl

Since we didn't get a response from the author for a long time, I pushed a bug-fix commit. The root cause is that nftables tables managed by Antrea not alway created. An error will be returned when trying to dump a non-existed table.

antoninbas reviewed (changes requested):
@hongliangl I am now reconsidering this approach. Why don't we go with the much simpler `nft list ruleset` command (equivalent to `iptables-save`)? This will be consistent with iptables and we will have all the rules, not just Antrea ones, which can be useful for troubleshooting. Only downside is that there will be duplicates when `iptables-nft` is used, but that's fine.

Something like this:
```go
dumpAllRules := func() error {
        output, err := d.executor.Command("nft", "list", "ruleset").CombinedOutput()
        // ...
}

if (d.v4Enabled && nftablesIPv4Supported()) || (d.v6Enabled && nftablesIPv6Supported()) {
        dumpAllRules()
}
```

antoninbas commented:
/test-all
